compiler/passes/checkResolved.cpp

```diff
@@ -76,6 +76,98 @@ static void checkForClassAssignOps(FnSymbol* fn) {
   }
 }
 
+// Returns the element type, given an array type.
+static Type* arrayElementType(AggregateType* arrayType) {
+  Type* eltType = NULL;
+  INT_ASSERT(arrayType->symbol->hasFlag(FLAG_ARRAY));
+  Type* instType = arrayType->getField("_instance")->type;
+  AggregateType* instClass = toAggregateType(canonicalClassType(instType));
+  TypeSymbol* ts = getDataClassType(instClass->symbol);
+  // if no eltType here, go to the super class
+  while(ts == NULL) {
+    if(Symbol* super = instClass->getSubstitutionWithName(astr("super"))) {
+        instClass = toAggregateType(canonicalClassType(super->type));
+        ts = getDataClassType(instClass->symbol);
+    } else break;
+  }
+  if(ts != NULL) eltType = ts->type;
+
+  return eltType;
+}
+
+// Returns the element type, given an array type.
+// Recurse into it if it is still an array.
+static Type* finalArrayElementType(AggregateType* arrayType) {
+  Type* eltType = NULL;
+  do {
+    eltType = arrayElementType(arrayType);
+    arrayType = toAggregateType(eltType);
+  } while (arrayType != NULL && arrayType->symbol->hasFlag(FLAG_ARRAY));
+
+  return eltType;
+}
+
+static bool isOrContains(Type *type, Flag flag, bool canBeTypeVar = false) {
+  if(type == NULL) return false;
+  else if(type->symbol->hasFlag(flag) ||
+          (canBeTypeVar && !type->symbol->hasFlag(FLAG_TYPE_VARIABLE))) {
```

Comment:
This is so code like this still works, because the field isn't actually a sync. Breaking this into a separate check and adding a comment
```chapel
record R {
  type t = sync int;
}
```